src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
-            return ref
-        raise AssertionError()
+    uid_gen_tmps = UIDGenerator(prefix="_tmp")
+
+    type_inference.infer_all(node, offset_provider=offset_provider, save_to_annex=True)
 
-    closures = []
+    tmps: list[ir.Sym] = []
+
+    closures: list[ir.StencilClosure] = []
     for closure in reversed(node.closures):
-        wrapped_stencil = ir.FunCall(fun=closure.stencil, args=closure.inputs)
-        popped_stencil = PopupTmps().visit(wrapped_stencil)
+        closure_stack: list[ir.StencilClosure] = [closure]
+        while closure_stack:
+            current_closure: ir.StencilClosure = closure_stack.pop()
+
+            if current_closure.stencil == im.ref("deref"):
+                closures.append(current_closure)
+                continue
+
+            # TODO(tehrengruber): extracting from a scan is untested.
+            is_scan: bool = isinstance(
+                current_closure.stencil, ir.FunCall
+            ) and current_closure.stencil.fun == im.ref("scan")
+            current_closure_stencil = (
+                current_closure.stencil if not is_scan else current_closure.stencil.args[0]  # type: ignore[attr-defined]  # ensured by is_scan
+            )
 
-        stencil_stack = [(closure.output, popped_stencil)]
-        domain = closure.domain
-        while stencil_stack:
-            output, call = stencil_stack.pop()
-            closure = ir.StencilClosure(
-                domain=domain,
-                stencil=call.fun,
-                output=output,
-                inputs=[handle_arg(arg) for arg in call.args],
+            stencil_body, extracted_lifts, _ = extract_subexpression(
+                current_closure_stencil.expr,
+                temporary_extraction_predicate,
+                uid_gen_tmps,
+                once_only=True,
+                deepest_expr_first=True,
             )
-            closures.append(closure)
-            domain = AUTO_DOMAIN
+
+            if extracted_lifts:
+                for tmp_sym, lift_expr in extracted_lifts.items():
+                    # make sure the applied lift is not capturing anything except of closure params
+                    used_symbol_refs = collect_symbol_refs(lift_expr)
+                    assert not (
+                        set(used_symbol_refs)
+                        - {param.id for param in current_closure_stencil.params}
+                    )
+
+                    assert isinstance(lift_expr, ir.FunCall) and isinstance(
+                        lift_expr.fun, ir.FunCall
+                    )
+
+                    # make sure the arguments to the applied lift are only symbols
+                    # (otherwise we would need to canonicalize using `canonicalize_applied_lift`
+                    # this doesn't seem to be necessary right now as we extract the lifts
+                    # in post-order of the tree)
+                    assert all(isinstance(arg, ir.SymRef) for arg in lift_expr.args)
```

Comment:
does it mean the `canonicalize_applied_lift` is currently not used? actually here is the comment (first line) that I want in the `canonicalize_applied_lift`.

Reply:
Yes. It looked too nice and untrivial to throw it away as I am not a 100% certain it is not needed. I can remove it if you like.